clients/client-common/src/main/java/org/eclipse/hono/client/util/CachingClientFactory.java

```diff
@@ -146,74 +173,139 @@ public void getOrCreateClient(
+                    ({} waiting creation requests for all keys)\
+                    """, key, waitingCreationRequests.size());
+            // this ensures that requests for a given key are completed in the order that the requests were made
+            waitingCreationRequests.add(new CreationRequestData(key, clientInstanceSupplier, result));
+            return;
+        }
 
         final T sender = activeClients.get(key);
-
         if (sender != null && livenessCheck.test(sender)) {
             log.debug("reusing cached client [{}]", key);
             result.handle(Future.succeededFuture(sender));
-        } else if (!creationLocks.computeIfAbsent(key, k -> Boolean.FALSE)) {
+            return;
+        }
+
+        if (creationLocks.putIfAbsent(key, Boolean.TRUE) == null) {
             // register a handler to be notified if the underlying connection to the server fails
             // so that we can fail the result handler passed in
             final Handler<Void> connectionFailureHandler = connectionLost -> {
                 // remove lock so that next attempt to open a sender doesn't fail
                 if (creationLocks.remove(key, Boolean.TRUE)) {
-                    result.handle(Future.failedFuture(
-                            new ServerErrorException(HttpURLConnection.HTTP_UNAVAILABLE, "no connection to service")));
+                    final var exception = createConnectionLostException();
+                    log.debug("failed to create new client for [{}]: {}", key, exception.toString());
+                    result.handle(Future.failedFuture(exception));
+                    failWaitingCreationRequests(key, exception);
                 } else {
                     log.debug("creation attempt already finished for [{}]", key);
                 }
             };
             creationRequests.add(connectionFailureHandler);
-            creationLocks.put(key, Boolean.TRUE);
             log.debug("creating new client for [{}]", key);
 
+            Future<T> clientInstanceSupplierFuture = null;
             try {
-                clientInstanceSupplier.get().onComplete(creationAttempt -> {
+                clientInstanceSupplierFuture = clientInstanceSupplier.get();
+            } catch (final Exception ex) {
+                creationLocks.remove(key);
+                creationRequests.remove(connectionFailureHandler);
+                log.error("exception creating new client for [{}]", key, ex);
+                activeClients.remove(key);
+                final ServerErrorException exception = new ServerErrorException(HttpURLConnection.HTTP_INTERNAL_ERROR,
+                        String.format("exception creating new client for [%s]: %s", key, ex.getMessage()));
+                result.handle(Future.failedFuture(exception));
+                failWaitingCreationRequests(key, exception);
+            }
+            if (clientInstanceSupplierFuture != null) {
```

Comment:
if we explicitly check for the supplier to actually return a Future, then I guess we should also explicitly handle the case when it doesn't, shouldn't we?

Reply:
I've added a `null` check for the returned Future now.